### DISCUSSION TITLE
GSoC 2022 Pedagogical Interactive Notebooks for Algorithms

### DISCUSSION THREAD

**Discussion by movahedi-ca** [networkx/networkx#5475] [Q&A] [answered]
Hi Guys,
My name is Mohammad, and I'm interested in working on interactive pedagogical notebooks for algorithms on NetworkX. I have all the requirements you asked for, and I'm already familiar with some of the simple algorithms. Also, I'm an inquisitive person and self-learner, so I'm confident I can do this project.

My question is, what are you guys looking for in the proposal for this project? Should I mention algorithms that I want to contribute toward them, or should I write about myself and my interest ? Can you please help with this?

Looking forward to hearing from you soon
Best regards,
Mohammad Hossein Movahedi
@rossbar @dschult  @MridulS

**Comment by MridulS** [comment, marked as answer]
Hi Mohammad,
Yes please do mention which algorithms you are interested in working on and a bit about your background. We also generally expect GSoC applicants to have made some sort of contribution to either NetworkX or the wider open source ecosystem in general. They need not be major fixes or changes, but it helps us to get a better overview of the applicant's background and skills.

You can also look at some applications from previous years at https://networkx.org/documentation/latest/developer/projects.html#completed-projects

Once you have created a draft of the applicant please do post the draft here, so other folks in the community can help review it and make it stronger :)

Cheers
Mridul

**Comment by movahedi-ca** [comment]
# GSoC Proposal

## NetworkX -Creating Pedagogical Interactive Notebooks for Algorithms

Mohammad Hossein Movahedi
[m.h.movahedi97@gmail.com](mailto:m.h.movahedi97@gmail.com)

April 2022

---

# Table of Content

- [GSoC Proposal](#gsoc-proposal)
  - [NetworkX -Creating Pedagogical Interactive Notebooks for Algorithms](#networkx--creating-pedagogical-interactive-notebooks-for-algorithms)
- [Table of Content](#table-of-content)
- [Student Details](#student-details)
  - [Personal Info](#personal-info)
  - [Education Info](#education-info)
    - [Bachelor Degree](#bachelor-degree)
    - [Master Degree](#master-degree)
  - [Programming Skills](#programming-skills)
  - [About Me](#about-me)
  - [Commitment to GSoC - Project](#commitment-to-gsoc---project)
- [Project Details](#project-details)
  - [Abstract](#abstract)
  - [The Plan](#the-plan)
  - [Project Timeline](#project-timeline)
    - [Period until accepted](#period-until-accepted)
    - [Phase 1 - Community Bonding Period](#phase-1---community-bonding-period)
    - [Phase 2 - Writing Two Notebooks](#phase-2---writing-two-notebooks)
    - [Phase 3 - Writing three more Notebooks](#phase-3---writing-three-more-notebooks)
    - [Phase 4 - Support and maintenance](#phase-4---support-and-maintenance)
- [Quick Q&A](#quick-qa)
- [References](#references)

---

# Student Details

## Personal Info

**Full Name**: Mohammad Hossein Movahedi

**Birth Date**: 21/October/1997

**Email**: [m.h.movahedi97@gmail.com](mailto:m.h.movahedi97@gmail.com)

**Github**: [https://github.com/momova97](https://github.com/momova97)

**Stack Overflow**: [https://stackoverflow.com/users/12489713](https://stackoverflow.com/users/12489713)

**LinkedIn** : [https://www.linkedin.com/in/mh-movahedi/](https://www.linkedin.com/in/mh-movahedi/)

**My website:** [https://www.mh-movahedi.com/](https://www.mh-movahedi.com/)

Phone Number: +

**Native Language**: Persian

**Location**: Toronto, Canada

**Timezone**: UTC - 5 (ECT time zone / Toronto)

## Education Info

### Bachelor Degree

Degree: Industrial Engineering

University: K. N. Toosi University of Technology

Final Grade: 15.2 / 20

year: 2015 - 2019

### Master Degree

Degree: Data analytics

University: Northeastern University

Current GPA : 3.7 / 4

year: 2022 - Ongoing

## Programming Skills

I started learning about algorithms when I was in grade ten. There has been something about algorithms that fascinates me since then. I studied some more advanced algorithms like genetic algorithms and other optimization algorithms during my undergraduate degree. I love digging into algorithms and trying to understand and deploy them.
The table below shows my skills in each field, and I determined where I have theoretical knowledge and where I have working experience.

| Category | Working Experience | Theoretical Knowledge  |
| --- | --- | --- |
| Programming languages | Python, C++, R, Matlab | VBS |
| Web development | HTML , CSS , Hugo static generator | NodeJs, Bootstrap, WP bakery  |
| Databases | MS Access, SQLite, | MySQL |
| Typesetting | Markdown | Latex |
| Version Control | Git |  |
| IDEs | Jupiter Notebook(IDE), Spyder, VS Code, Sublime, R-Studio |  |
| Algorithms  | Genetic, Neural network, Fuzzy theory, | Data analytics, Graph theories, optimization algorithms  |

## About Me

I see areas of Improvement wherever I look. I spend most of my time discovering how to enhance things. I’m passionate about solving abstract issues and developing creative ways to crack them.
I love algorithms. I want to understand them, improve them and deploy them.

I like studying and learning about abstract ideas. My favorite topic has always been philosophy because instead of asking “How,” philosophy asks “Why.” I believe the why is more impotent than the How. and that’s one of the reasons I am attracted to programming and algorithms, to be specific. I like to know what’s happening behind the hood instead of just driving the car, and I love sharing my findings with others.

## Commitment to GSoC - Project

I’m a student at northeastern university, and my objective in doing this project is to create a portfolio for my resume and contribute to open-source projects.

As a student, I have lots of time to devote to this project and advance it during the summer. Based on my ambitious and persistent personality, interest in Algorithms, and background and knowledge of python programming and graph theories, I'm sure I can deliver this project on time and within the expected scope.

This project is the best thing that I can do this summer to have fun and create a network of amazing people who genuinely believe in Linus Torvalds, Christine Peterson, Eric Raymond, and Bruce Perens’ vision of open-source and, more importantly, as a member of the open-source community, contribute towards it. 

---

# Project Details

## Abstract

NetworkX is a Python library that intends to be the reference library for network science methods, emphasizing graph data structures and algorithms for network science applications. The organization prioritizes easy-to-understand implementations over maximizing performance, education, and documentation.

Even though the algorithms are fully documented, descriptions of the concepts behind the algorithms are frequently lacking, demanding a need to clarify these concepts and analyze the algorithms.

## The Plan

This summer, I chose to do the large project mainly because I will be free for the entire summer, and I know I can do it in time based on the following timeline.

As I mentioned earlier, I love working with algorithms, and I’m not afraid of dealing with hard-to-understanding ones. I would love to understand and create notebooks and examples for hard ones. Thus, they are the ones that might need an example the most.

To maximize the effect of my help to the community of NetworkX algorithms users, I choose to work on five algorithms with different approaches that there is no example for them on the website.

the below list shows algorithms I like to work on them during this project

  

| Name Of Algorithms | link |
| --- | --- |
| Boundary | https://networkx.org/documentation/latest/reference/algorithms/boundary.html |
| Chains | https://networkx.org/documentation/latest/reference/algorithms/chains.html |
| Clique | https://networkx.org/documentation/latest/reference/algorithms/clique.html |
| Cores | https://networkx.org/documentation/stable/reference/algorithms/core.html |
| Cuts | https://networkx.org/documentation/stable/reference/algorithms/cuts.html |

## Project Timeline

[Project Timeline](https://www.notion.so/mh-movahedi/b47b0d53c2f34e97a468d978b59ca948?v=1b1463de94dd4fdeafd28467ca18f387)

### Period until accepted

During this period, I will stay in touch with primary mentors, try to familiarize myself with the source code, watch the discussion boards regularly, and be an active member of the community here at NetworkX. Also, I will revisit my notes about graph theories and learn more about NetworkX.

### Phase 1 - Community Bonding Period

During this phase, I will communicate even more than in the first part with mentors. With their help, I will start a poll to ask users about which five algorithms they want a notebook for because I chose mentioned algorithms based lack of information about them on the website. Still, maybe some popular algorithms are well documented, but users feel that they need more examples and better documents for them. In this part, with the community's help, we can pin down those algorithms so my contribution can be more productive. 

### Phase 2 - Writing Two Notebooks

during this phase, I will do the actual work and start creating the Jupiter notebook for the chosen algorithms. I decide to do two books in this Phase and three in the next phase because I worry there might be some unforeseen difficulty in creating notebooks. It may cause the first notebook to take more time to be written, but once I get the hang of the work for the next ones, I will finish the notebooks faster and more efficiently.

### Phase 3 - Writing three more Notebooks

In this Phase, I will write three more notebooks, address feedback on the previous phase if necessary and try to get public attention to notebooks to get even more feedback on them and try to improve them even further by the end of the project.

### Phase 4 - Support and maintenance

I believe everything eventually needs support and maintenance, even a notebook. Hopefully, after doing this project, I will continue my work on other algorithms that we detected in phase 1 and continue to maintain the notebooks if any change happens to the codes and algorithms.

---

# Quick Q&A

1. Have you communicated with the organization's mentors?
yes, I have, and I will continue making better communications
2. Have you communicated with the community?
I posted a question but nothing more than that.
3. Did you reference projects you coded WITH links to repos or provided code?
the repo of my website, which is generated by Hugo (open-source) and based on an excellent open-source template and is built on Netlify (open-source), is provided below
    
    [https://github.com/momova97/f9.git](https://github.com/momova97/f9.git)
    
4. Did you provide several methods to contact you? (email, skype, mobile/phone, Twitter, chat, and/or Tumblr if available)
yes, they can be found [Personal Info](#personal-info)
5. Did you include a preliminary project plan (before, during, after GSoC)?
yes, it can be found [Project Timeline](#project-timeline)
6. Did you state which project you are applying for and why you think you will end up completing the project?
yes, the answer to this question can be found in [About me](#about-me)  and also in [the plan](#the-plan)
7. Do you have time for GSoC? This is a paid job! State that you have time in your motivation letter and list other commitments!
yes, I’m free all the summer, and my motivation to do this project can be found in [Commitment to GSoC - Project](#commitment-to-gsoc---project)
8. Did you add a link to ALL your application files to a cloud hoster like GitHub or Dropbox? (easy points! wink)
yes, It’s all online
9. Did you create a pull request on the existing code?
not yet but I’m planning on creating one based on the [instruction provided on the website](https://networkx.org/documentation/latest/developer/contribute.html)
10. Do you continue communication until accepted students are announced?
    
    I will
    

---

# References

[Contributor Guide - NetworkX 2.8 documentation](https://networkx.org/documentation/latest/developer/contribute.html)

[https://github.com/momova97/f9](https://github.com/momova97/f9)

**Reply by movahedi-ca** [reply]
@rossbar @dschult @MridulS
Hi Guys,
could you please look at the proposal I shared here and give me some comments on it?
Best Regards,
Mohammad

**Reply by MridulS** [reply]
All looks good @momova97! Please do submit the proposal on the official GSoC website before the deadline :)

**Reply by movahedi-ca** [reply]
Thank you,
By the way, I suggest that we create a poll so users can tell which algorithms they want to see examples for, and then I develop notebooks for the five most wanted ones ? is it a good idea ? is it possible?

**Reply by MridulS** [reply]
I like the idea but unfortunately we don't really have a way of reaching all of our users, github and mailing list is visited by only a small number of networkx users.

**Reply by movahedi-ca** [reply]
I understand :(
I hoped there is a way we could do that, like reaching out to communities. Besides our community on Github, do you have any community centered on network algorithms? maybe I can reach them and ask them about their suggestion. Perhaps this even could lead to strengthing the community here.

Also, is there any way for me to assist the NetworkX until I start my project?

**Reply by MridulS** [reply]
> community centered on network algorithms?

Not that I am aware of :/

> Also, is there any way for me to assist the NetworkX until I start my project?

Feel free to contribute in any way you would like :) If you want to fix issues please go through our [issue tracker](https://github.com/networkx/networkx/issues), if you would like to directly work on the nx-guides notebooks feel free to start working on it (but do keep in mind that this would be independent of the outcome of GSoC :) ), if you would like to implement new algorithms start a discussion about it!

**Reply by movahedi-ca** [reply]
Thank you for your prompt response,
I will start creating an nx-guides notebook independent of GSoC.I think in this way, I can effectively use my time to get adjusted to notebooks.
